examples/basecoin/app/app.go

```diff
@@ -1,95 +1,121 @@
 package app
 
 import (
+	"encoding/json"
 	"fmt"
-	"os"
 
 	bam "github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/examples/basecoin/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	"github.com/cosmos/cosmos-sdk/x/bank"
+	"github.com/cosmos/cosmos-sdk/x/sketchy"
 
-	"github.com/tendermint/abci/server"
-	abci "github.com/tendermint/abci/types"
+	crypto "github.com/tendermint/go-crypto"
 	"github.com/tendermint/go-wire"
 	cmn "github.com/tendermint/tmlibs/common"
 )
 
-const appName = "BasecoinApp"
-
-// BasecoinApp - extended ABCI application
+// Extended ABCI application
 type BasecoinApp struct {
 	*bam.BaseApp
-	router     bam.Router
-	cdc        *wire.Codec
-	multiStore sdk.CommitMultiStore //TODO distinguish this store from *bam.BaseApp.cms <- is this one master?? confused
+	cdc *wire.Codec
 
-	// The key to access the substores.
+	// keys to access the substores
 	capKeyMainStore *sdk.KVStoreKey
 	capKeyIBCStore  *sdk.KVStoreKey
 
-	// Object mappers:
+	// Manage getting and setting accounts
 	accountMapper sdk.AccountMapper
 }
 
-// NewBasecoinApp - create new BasecoinApp
-// TODO: This should take in more configuration options.
-// TODO: This should be moved into baseapp to isolate complexity
 func NewBasecoinApp(genesisPath string) *BasecoinApp {
 
-	// Create and configure app.
-	var app = &BasecoinApp{}
-
-	// TODO open up out of functions, or introduce clarity,
-	// interdependancies are a nightmare to debug
-	app.initCapKeys() // ./init_capkeys.go
-	app.initBaseApp() // ./init_baseapp.go
-	app.initStores()  // ./init_stores.go
-	app.initBaseAppInitStater()
-	app.initHandlers() // ./init_handlers.go
-
-	genesisiDoc, err := bam.GenesisDocFromFile(genesisPath)
-	if err != nil {
-		panic(fmt.Errorf("error loading genesis state: %v", err))
+	// create your application object
+	var app = &BasecoinApp{
+		BaseApp:         bam.NewBaseApp("BasecoinApp"),
+		cdc:             MakeTxCodec(),
+		capKeyMainStore: sdk.NewKVStoreKey("main"),
+		capKeyIBCStore:  sdk.NewKVStoreKey("ibc"),
 	}
 
-	// set up the cache store for ctx, get ctx
-	// TODO: can InitChain handle this too ?
-	app.BaseApp.BeginBlock(abci.RequestBeginBlock{Header: abci.Header{}})
-	ctx := app.BaseApp.NewContext(false, nil) // context for DeliverTx
-
-	// TODO: combine with InitChain and let tendermint invoke it.
-	err = app.BaseApp.InitStater(ctx, genesisiDoc.AppState)
+	// define the accountMapper
+	app.accountMapper = auth.NewAccountMapperSealed(
+		app.capKeyMainStore, // target store
+		&types.AppAccount{}, // prototype
+	)
+
+	// add handlers
+	app.Router().AddRoute("bank", bank.NewHandler(bank.NewCoinKeeper(app.accountMapper)))
```

Comment:
I feel like the coinkeeper should be set to app first.

Reply:
Is this what you meant? 
``` golang
coinKeeper := bank.NewCoinKeeper(app.accountMapper)
app.Router().AddRoute("bank", bank.NewHandler(coinKeeper))
```